#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

## Librerias

In [218]:
import pandas as pd

import urllib.request
from PIL import Image

## Data

In [221]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("data/train.csv")


## Data exploration

In [222]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 912 entries, 0 to 911
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         912 non-null    int64  
 1   Company           912 non-null    object 
 2   Product           912 non-null    object 
 3   TypeName          912 non-null    object 
 4   Inches            912 non-null    float64
 5   ScreenResolution  912 non-null    object 
 6   Cpu               912 non-null    object 
 7   Ram               912 non-null    object 
 8   Memory            912 non-null    object 
 9   Gpu               912 non-null    object 
 10  OpSys             912 non-null    object 
 11  Weight            912 non-null    object 
 12  Price_euros       912 non-null    float64
dtypes: float64(2), int64(1), object(10)
memory usage: 92.8+ KB


In [223]:
train.head()

laptop_ID Company            Product   TypeName  Inches   
0       1104    Acer     Aspire ES1-523   Notebook    15.6  \
1        114    Dell             XPS 13  Ultrabook    13.3   
2        172    Acer    Aspire A517-51G   Notebook    17.3   
3        918      HP      Elitebook 820    Netbook    12.5   
4        447  Lenovo  IdeaPad 320-15AST   Notebook    17.3   

                   ScreenResolution                         Cpu  Ram   
0                          1366x768   AMD A8-Series 7410 2.2GHz  4GB  \
1  Quad HD+ / Touchscreen 3200x1800  Intel Core i7 7560U 2.4GHz  8GB   
2       IPS Panel Full HD 1920x1080  Intel Core i5 8250U 1.6GHz  8GB   
3                 Full HD 1920x1080  Intel Core i7 7500U 2.7GHz  8GB   
4                          1600x900   AMD A6-Series 9220 2.5GHz  8GB   

      Memory                           Gpu       OpSys  Weight  Price_euros  
0  500GB HDD                 AMD Radeon R5  Windows 10   2.4kg        387.0  
1  256GB SSD  Intel Iris Plus Graphics 640  Windows 10  1.23kg       1379.0  
2  256GB SSD          Nvidia GeForce MX150  Windows 10     3kg        854.0  
3  512GB SSD         Intel HD Graphics 620  Windows 10  1.26kg       1483.0  
4    1TB HDD                 AMD Radeon R4  Windows 10   2.8kg        519.0

## Data processing

In [228]:

ordenador = ["Ram", "Weight", "Price_euros"]
df = train[ordenador]


In [229]:
df.dropna(inplace=True)


C:\Users\horta\AppData\Local\Temp\ipykernel_1360\658506679.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.dropna(inplace=True)


In [230]:
import re
df['Ram'] = df['Ram'].apply(lambda x: re.sub(r'\D+', '', x) if isinstance(x, str) else x)
df['Ram'] = pd.to_numeric(df['Ram'])

C:\Users\horta\AppData\Local\Temp\ipykernel_1360\3402529933.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Ram'] = df['Ram'].apply(lambda x: re.sub(r'\D+', '', x) if isinstance(x, str) else x)
C:\Users\horta\AppData\Local\Temp\ipykernel_1360\3402529933.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Ram'] = pd.to_numeric(df['Ram'])


In [227]:

df['OpSys'] = df['OpSys'].str.lower()

df['OpSys'] = df['OpSys'].replace({'mac os x': 'macos', 'windows 7': 'windows 10'})

df['OpSys'] = df['OpSys'].astype('category')

KeyError: 'OpSys'

In [231]:
df["Weight"]=df.Weight.str.replace("[^0-9]","",regex=True)

C:\Users\horta\AppData\Local\Temp\ipykernel_1360\3873500035.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Weight"]=df.Weight.str.replace("[^0-9]","",regex=True)


In [232]:
df.head()

Ram Weight  Price_euros
0    4     24        387.0
1    8    123       1379.0
2    8      3        854.0
3    8    126       1483.0
4    8     28        519.0

In [233]:
df['Weight'] = pd.to_numeric(df['Weight'], errors='coerce')

df['Weight'] = df['Weight'].apply(lambda x: x/1000 if x > 1000 else x)

C:\Users\horta\AppData\Local\Temp\ipykernel_1360\3946339263.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Weight'] = pd.to_numeric(df['Weight'], errors='coerce')
C:\Users\horta\AppData\Local\Temp\ipykernel_1360\3946339263.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Weight'] = df['Weight'].apply(lambda x: x/1000 if x > 1000 else x)


In [234]:
df['Weight'].head()

0     24.0
1    123.0
2      3.0
3    126.0
4     28.0
Name: Weight, dtype: float64

In [235]:
df['Price_euros'] = df['Price_euros'].astype(float)

C:\Users\horta\AppData\Local\Temp\ipykernel_1360\2596229545.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Price_euros'] = df['Price_euros'].astype(float)


In [236]:
df.head()

Ram  Weight  Price_euros
0    4    24.0        387.0
1    8   123.0       1379.0
2    8     3.0        854.0
3    8   126.0       1483.0
4    8    28.0        519.0

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [238]:
X = df[['Ram', 'Weight']]
y = df['Price_euros']


In [239]:
X.head()

Ram  Weight
0    4    24.0
1    8   123.0
2    8     3.0
3    8   126.0
4    8    28.0

### 2. Dividir X_train, X_test, y_train, y_test

In [240]:
from sklearn.model_selection import train_test_split

In [241]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [242]:
X_train

Ram  Weight
25     4    23.0
84    16   272.0
10    16    25.0
342    8   190.0
890   16   129.0
..   ...     ...
106    8     2.0
270   16    25.0
860    4    28.0
435    6    23.0
102    4    21.0

[729 rows x 2 columns]

### 3. Asignar el modelo (vacío) a una variable


In [243]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [244]:
model.fit(X_train,y_train)

LinearRegression()

In [246]:
pred = model. predict(X_test)

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

In [247]:
from sklearn.metrics import mean_squared_error
import numpy as np
print("RMSE: ", np.sqrt(mean_squared_error(y_test, pred)))

RMSE:  486.585822878403


-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

In [ ]:
#No entiendo muy bien esta parte

### 2. Carga los datos de `test.csv` para predecir.

In [274]:
X_pred = pd.read_csv("data/test.csv")
X_pred.head()

laptop_ID   Company          Product            TypeName  Inches   
0        750   Toshiba  Tecra X40-D-10G            Notebook    14.0  \
1        726  Mediacom    Smartbook 142            Notebook    14.0   
2        633  Mediacom    SmartBook 140            Notebook    14.0   
3        363      Dell    Inspiron 7577              Gaming    15.6   
4        319      Asus    VivoBook Flip  2 in 1 Convertible    11.6   

                            ScreenResolution   
0  IPS Panel Full HD / Touchscreen 1920x1080  \
1                IPS Panel Full HD 1920x1200   
2                          Full HD 1920x1080   
3                          Full HD 1920x1080   
4                       Touchscreen 1366x768   

                                    Cpu  Ram              Memory   
0            Intel Core i5 7200U 2.5GHz  8GB           256GB SSD  \
1              Intel Atom Z8350 1.92GHz  4GB            32GB HDD   
2           Intel Atom x5-Z8350 1.44GHz  2GB  32GB Flash Storage   
3           Intel Core i5 7300HQ 2.5GHz  8GB             1TB HDD   
4  Intel Celeron Dual Core N3350 1.1GHz  2GB  32GB Flash Storage   

                       Gpu       OpSys  Weight  
0    Intel HD Graphics 620  Windows 10  1.25kg  
1    Intel HD Graphics 400  Windows 10   1.4kg  
2        Intel HD Graphics  Windows 10   1.4kg  
3  Nvidia GeForce GTX 1050  Windows 10  2.65kg  
4    Intel HD Graphics 500  Windows 10   1.1kg

In [275]:
X_pred.dropna(inplace=True)


In [276]:
import re
X_pred['Ram'] = X_pred['Ram'].apply(lambda x: re.sub(r'\D+', '', x) if isinstance(x, str) else x)
X_pred['Ram'] = pd.to_numeric(X_pred['Ram'])


In [277]:
X_pred["Weight"]=X_pred.Weight.str.replace("[^0-9]","",regex=True)

In [278]:

X_pred['Weight'] = X_pred['Weight'].str.replace("[^0-9]", "", regex=True)
X_pred['Weight'] = pd.to_numeric(X_pred['Weight'], errors='coerce')
X_pred['Weight'] = X_pred['Weight'].apply(lambda x: x/1000 if x > 1000 else x)


In [279]:
X_pred =X_pred[["Ram", "Weight"]]

In [280]:
X_pred.columns

Index(['Ram', 'Weight'], dtype='object')

In [281]:
X_pred.shape

(391, 2)

**¿Por qué da error?** 

In [282]:
predictions_submit = model.predict(X_pred)
predictions_submit

array([1092.02150804,  623.87862076,  417.84522111, 1162.74810553,
        416.32965117,  627.92014061,  617.81634097, 2015.17234309,
       1865.13091843, 1041.50250984, 1122.83809695, 1111.21872736,
       1040.49212988, 1039.9869399 ,  713.29724757, 1121.82771698,
       1090.00074812, 1951.51840535,  627.92014061, 1039.9869399 ,
       1037.46098999, 1164.76886545,  732.49446689, 1125.36404686,
        710.77129766, 1087.97998819,  933.47596671, 1043.01807979,
        627.92014061, 1041.50250984, 1130.92113666, 1029.88314026,
        834.96392022, 1040.99731986, 1038.97655993, 1081.9177084 ,
       1038.97655993, 1036.9558    , 1126.8796168 , 1921.71219641,
       1874.72952809, 1041.50250984,  690.0585084 , 1098.08378783,
       1039.48174991, 2001.53221357, 1152.64430589, 1854.52192881,
       1161.73772556, 1992.94398388,  733.50484685,  730.47370696,
       1142.03531626, 1123.84847691,  627.92014061,  719.86471734,
       1040.99731986,  679.95470876, 1096.56821788, 1036.45061

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

In [ ]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [283]:
predictions_submit = model.predict(X_pred[["Ram", "Weight"]])
predictions_submit

array([1092.02150804,  623.87862076,  417.84522111, 1162.74810553,
        416.32965117,  627.92014061,  617.81634097, 2015.17234309,
       1865.13091843, 1041.50250984, 1122.83809695, 1111.21872736,
       1040.49212988, 1039.9869399 ,  713.29724757, 1121.82771698,
       1090.00074812, 1951.51840535,  627.92014061, 1039.9869399 ,
       1037.46098999, 1164.76886545,  732.49446689, 1125.36404686,
        710.77129766, 1087.97998819,  933.47596671, 1043.01807979,
        627.92014061, 1041.50250984, 1130.92113666, 1029.88314026,
        834.96392022, 1040.99731986, 1038.97655993, 1081.9177084 ,
       1038.97655993, 1036.9558    , 1126.8796168 , 1921.71219641,
       1874.72952809, 1041.50250984,  690.0585084 , 1098.08378783,
       1039.48174991, 2001.53221357, 1152.64430589, 1854.52192881,
       1161.73772556, 1992.94398388,  733.50484685,  730.47370696,
       1142.03531626, 1123.84847691,  627.92014061,  719.86471734,
       1040.99731986,  679.95470876, 1096.56821788, 1036.45061

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [284]:
sample = pd.read_csv("data/sample_submission.csv")

In [285]:
sample.head()

laptop_ID  Price_euros
0        750          500
1        726          500
2        633          500
3        363          650
4        319          650

In [286]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [287]:
submission = pd.DataFrame({"laptop_ID": sample["laptop_ID"], "Price_euros": predictions_submit}) 

In [288]:
submission.head()

laptop_ID  Price_euros
0        750  1092.021508
1        726   623.878621
2        633   417.845221
3        363  1162.748106
4        319   416.329651

In [289]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [291]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.laptop_ID.all() == sample.laptop_ID.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [292]:
chequeator(submission)

You're ready to submit!
